### DISCUSSION TITLE
Why is the docker image so large?

### DISCUSSION THREAD

**Discussion by chemwolf6922** [open-webui/pipelines#282] [General]
The pipelines' docker image is currently sitting at 4.37GB. This is far too large for its functionalities. 
This may be fine on desktops with TB of storage. But it is making deploying this with the web-ui on SBCs harder. (When updating, two versions will consume over 20GB). 
Please help investigate why this is the case. Are all the python dependencies really needed for this project? 
Thanks.

**Comment by chemwolf6922** [comment]
I read the code for a little. I think the framework should only require very little dependencies. As for the pipelines. I'll suggest let the pipeline provide a requirements.txt with the .py file. This should increase the extensibility while reducing the default install size. 
Also, the largest package pytorch does not seem to be used in the code (according to a quick search). I think it might be a good idea to do a purge on the dependencies.

**Comment by jtha** [comment]
Is it as simple as removing all of the ML libraries from the requirements.txt used in the main build? If it breaks things, then would to set up environment variable and conditionals everywhere those libraries are referenced. I'd be up for doing this in a new branch or fork so we have a super lightweight image to start off with.

**Reply by jtha** [reply]
Update:

Yes, it is that simple. Forked it, removed every python library not directly referenced in any of the .py files of the repo from requirements.txt and removed the pytorch install from the Dockerfile and just simply build the images. 183MB compressed (pulled from hub), 641MB image (compared to 2.1GB for original image). 

Can confirm all current pipelines I'm using are running fine, including ones where additional libraries are needed (specified in the respective .py docstrings at the top).

This is the final requirements.txt I used:
```fastapi==0.111.0
uvicorn[standard]==0.22.0
pydantic==2.7.1
python-multipart==0.0.9
python-socketio
grpcio

passlib==1.7.4
passlib[bcrypt]
PyJWT[crypto]

requests==2.32.2
aiohttp==3.9.5
httpx

**Reply by chemwolf6922** [reply]
Thanks for the testing. I was planning to do the same. Have you tried removing the intermediate files like this PR https://github.com/open-webui/pipelines/pull/279? 
Also, would you like to share the improvements by creating a pull request?

**Reply by chemwolf6922** [reply]
I tried the fixes in the PR. (Also removed the torch packages). I managed to reduce the image size down to 251MB (x86)

**Reply by chemwolf6922** [reply]
We should definitely create a PR on this. Should we add a parameter like minimum build? And maybe add new image tags x.x-minimum?

**Comment by chemwolf6922** [comment]
I created a PR here https://github.com/open-webui/pipelines/pull/300 .
I did not apply changes in https://github.com/open-webui/pipelines/pull/279 as those intermediate packages might still be useful for installing dependencies on the fly.
